src/Opaleye/WithRecursive.hs

```diff
@@ -0,0 +1,78 @@
+{-# LANGUAGE FlexibleContexts #-}
+
+module Opaleye.WithRecursive
+  ( withRecursive
+  , withRecursiveExplicit
+  )
+where
+
+import Data.List.NonEmpty (NonEmpty((:|)))
+import Data.Profunctor.Product.Default (Default, def)
+import Opaleye.Internal.Binary (Binaryspec(..), runBinaryspec)
+import Opaleye.Internal.QueryArr (Select, SelectArr, runSimpleQueryArr', productQueryArr')
+import Opaleye.Internal.Unpackspec (Unpackspec(..), runUnpackspec)
+import Opaleye.Internal.PackMap (PackMap(..))
+import qualified Opaleye.Internal.HaskellDB.PrimQuery as HPQ
+import qualified Opaleye.Internal.PrimQuery           as PQ
+import qualified Opaleye.Internal.Sql                 as Sql
+import qualified Opaleye.Internal.Tag                 as Tag
+import qualified Opaleye.Internal.PackMap as PM
+
+
+withRecursive :: Default Binaryspec a a => Select a -> SelectArr a a -> Select a
```

Comment:
Do you know if this can be generalised to `Select a -> (a -> Select a) -> Select a`? If it can it would be good to do so. If it can't it would be good to document why.

Given that there's a `LATERAL` perhaps it can. I'd also like to know to what extent it depends on the `LATERAL`, perhaps having another version without the `LATERAL` that uses `addTable` (although I don't understand `addTable`).

Reply:
You're right that the type can be generalised to `Select a -> (a -> Select) -> Select a`, so I've done that.

Reply:
As for another version that uses `addTable`, I don't fully understand `addTable` either, and the comments on the previous implementation implied that it was incomplete. As for the extent to which it uses `LATERAL`, maybe I'll just show the SQL it generates:

```sql
SELECT
CAST("with0_8" AS int8) as "anon"
FROM (SELECT
      *
      FROM (WITH RECURSIVE "cte_1" AS
            (SELECT
             "values0_2" as "binary0_3"
             FROM (SELECT
                   *
                   FROM (SELECT "column1" as "values0_2"
                         FROM
                         (VALUES
                          (CAST(1 AS int8)),
                          (CAST(2 AS int8)),
                          (CAST(3 AS int8)),
                          (CAST(5 AS int8))) as "V") as "T1") as "T1"
             UNION ALL
             SELECT
             "values0_5" as "binary0_3"
             FROM (SELECT
                   *
                   FROM (SELECT
                         "binary0_3" as "rec0_4"
                         FROM "cte_1" as "T1") as "T1",
                        LATERAL
                        (SELECT "column1" as "values0_5"
                         FROM
                         (VALUES
                          (("rec0_4") * (CAST(2 AS int8))),
                          (("rec0_4") * (CAST(3 AS int8)))) as "V") as "T2"
                   WHERE (("values0_5") < (CAST(10 AS int8)))) as "T1")
            SELECT
            "binary0_3" as "with0_8"
            FROM "cte_1" as "T1") as "T1") as "T1"
```

(That's generated by the Rel8 code:
```haskell
putStrLn $ showQuery $ loop (values [1 :: Expr Data.Int.Int64, 2, 3, 5]) (\a -> do x <- values [a * 2, a * 3]; where_ (x <. 10); pure x)
```
but the Opaleye would be very similar)

So, the general form is:

```sql
WITH RECURSIVE "cte_N" AS
      (SELECT
       [...] as "binaryN_N"
       FROM (...) as "T1" -- base case
       UNION ALL
       SELECT
       [...] as "binaryN_N"
       FROM (SELECT
             *
             FROM (SELECT
                   "binaryN_N" as "recN_N"
                   FROM "cte_1" as "T1") as "T1",
                  LATERAL
                  (...) as "T2") as "T1") -- recursive case, from inside which
                                          -- we want to be able to refer to
                                          -- columns like recN_N
      SELECT
      "binaryN_N" as "withN_N"
      FROM "cte_N" as "T1"
```